org.adempiere.base/src/org/compiere/model/MRole.java

```diff
@@ -3428,4 +3441,31 @@ public MRole markImmutable() {
 		return this;
 	}
 
+	/**
+	 * CCache listener class to reload the default role 
+	 */
+	private static class CCacheListener extends CCache<String, Object> {
+		/**
+		 * 
+		 */
+		private static final long serialVersionUID = 2726872291294866031L;
+
+		protected CCacheListener(String tableName) {
+			super(tableName, tableName+"|CCacheListener", 0, false);
+		}
+
+		public int reset() {
+			getDefault(Env.getCtx(), true); // reload the default role
+			return super.reset();
+		}
+
+		@Override
+		public int reset(int recordId) {
+			if (recordId == Env.getAD_Role_ID(Env.getCtx()))
+				getDefault(Env.getCtx(), true); // reload the default role
+			return super.reset(recordId);
+		}
+
+	}
```

Comment:
I don't think we should use cache listener to reload MRole.getDefault since that's not using cache at all. Should just call mrole.getdefault from the model class.

Reply:
The right approach should be through using the IMessageService instead (Core.getMessageService()). The web client, upon login should subscribe to the "Reload Role" message topic, keep a reference to the desktop and when message arrive for that topic, schedule a event (Executions.schedule) to reload the role. In MRole's afterSave and beforeSave method, you will publish the "Reload Role" message to the "Reload Role" topic.

Reply:
Another option is to address this is to add a loaded role cache and in the getDefaultRole() method, after getting role from Env.getCtx(), check that the role exists in the loaded role cache and return null if it is not. This way, role will be reloaded after reset of cache.

Reply:
Thanks @hengsin 

I started implementing your second suggestion (check if the role in context is in the cache), and then I found that it is a very easy task to move the default role to be a cache instead of keeping it in context.

When moved the default role to be a CCache, everything worked without need of special tricks.

Reply:
Thanks @hengsin - implemented your suggestion, tested and it works perfect.